Task 1 (Creating main page - header, footer, cities list + filtering)

StefanAtanasov98 opened the pull request:
Creating the main page of Weather App includes the following components/functionalities:

- Header with the name of the app;
- Footer with the current time;
- Navigation bar with a dropdown button containing different cities and an input field for filtering them;

<img width="960" alt="app" src="https://user-images.githubusercontent.com/103125976/192246445-f5d8e5a9-ac9a-4af3-b198-ff984fb8dd8c.PNG">

StefanAtanasov98 committed:
Header, footer and dropdown added (basic styles)

virzhinia-petkova-redshelf commented on src/components/Homepage/Homepage.js:
```diff
@@ -0,0 +1,37 @@
+import * as homepageStyles from '../../styles/homepage.module.css';
+import { Nav } from './Nav/Nav.js';
+
+export function Homepage(props) {
```
1. it's much better not to export the function directly. Define it first and then export it at the bottom of the file. If you don't do that you'll have a hard time adding prop types. Read more about it here https://reactjs.org/docs/typechecking-with-proptypes.html#gatsby-focus-wrapper
2. The convention for naming react components is PascalCase aka HomePage

virzhinia-petkova-redshelf commented on src/components/Homepage/Homepage.js:
```diff
@@ -0,0 +1,37 @@
+import { Nav } from './Nav/Nav.js';
+
+export function Homepage(props) {
+  const [time, setTime] = useState('Loading...');
```
If you'd like to have loading show up, use a different state instance. It's weird to change the data you have for your state with something unrelated - think of it as what is the value of time, not what the UI should render. You can use null for initial value here.

virzhinia-petkova-redshelf commented on src/components/Homepage/Homepage.js:
```diff
@@ -0,0 +1,37 @@
+    const date = new Date();
+    const hours = date.getHours() < 10 ? `0${date.getHours()}` : date.getHours();
+    const minutes = date.getMinutes() < 10 ? `0${date.getMinutes()}` : date.getMinutes();
+    const seconds = date.getSeconds() < 10 ? `0${date.getSeconds()}` : date.getSeconds();
```
all of those looks very similar - can you create a reusable function for them?

virzhinia-petkova-redshelf commented on src/components/Homepage/Homepage.js:
```diff
@@ -0,0 +1,37 @@
+import React, { useState, useEffect } from 'react';
+import * as homepageStyles from '../../styles/homepage.module.css';
```
please use camelCase for such things aka homePage...

virzhinia-petkova-redshelf commented on src/components/Homepage/Homepage.js:
```diff
@@ -0,0 +1,37 @@
+
+  return (
+    <>
+      <header>
```
Is this page the only place you're going to use the header? What about the feedback/info page? Wouldn't you like to use it there as well? 

It will be a far better to use it as a separate component. The footer too. I'd suggest you think about structure/components and see if you can come up with a more robust structure.

virzhinia-petkova-redshelf commented on src/components/Homepage/Homepage.js:
```diff
@@ -0,0 +1,37 @@
+  return (
+    <>
+      <header>
+        <h1>Weather Forecast</h1>
```
move this to a constant please

StefanAtanasov98 replied:
Could you elaborate on this? I assume we want it to be constant so that we can manipulate it safely/easily in the future or..?

virzhinia-petkova-redshelf replied:
You can refer to this conversation as it's discussing the same question - https://github.com/ximet/ReactTask/pull/290/files#r975016810

virzhinia-petkova-redshelf commented on src/components/Homepage/Homepage.js:
```diff
@@ -0,0 +1,37 @@
+      </header>
+      <section className="layout">
+        <Nav styles={homepageStyles} />
+        <main></main>
```
why do you have an empty main? Also main should be the one the wraps all components in the main content(this does not include header/footer/etc), see https://developer.mozilla.org/en-US/docs/Web/HTML/Element/main

virzhinia-petkova-redshelf commented on src/components/Homepage/Homepage.js:
```diff
@@ -0,0 +1,37 @@
+        <main></main>
+      </section>
+      <footer>
+        <span>Current time: {time}</span>
```
constant please

virzhinia-petkova-redshelf commented on src/components/Homepage/Nav/DropdownMenu.js:
```diff
@@ -0,0 +1,35 @@
+  const [isOpen, setIsOpen] = useState(false);
+
+  const showCities = ev => {
+    ev.preventDefault();
```
you don't need to use preventDefault() since this is just a button

virzhinia-petkova-redshelf commented on src/components/Homepage/Nav/DropdownMenu.js:
```diff
@@ -0,0 +1,35 @@
+    const dropdownContent = [...document.getElementsByClassName(styles.dropdownContent)][0];
+    if (isOpen) {
+      setIsOpen(false);
+      dropdownContent.style.display = 'none';
```
avoid altering css from js if you can help it

virzhinia-petkova-redshelf replied:
Can you use a condition in JSX instead of adjusting the css? Something like(pseudo code):
`{isOpen ? dropdownContent : null}`

virzhinia-petkova-redshelf commented on src/components/Homepage/Nav/DropdownMenu.js:
```diff
@@ -0,0 +1,35 @@
+  return (
+    <>
+      <div className={styles.dropdown}>
+        <button className={styles.dropdownBtn} onClick={showCities}>
```
do you have an input field for filtering?

virzhinia-petkova-redshelf commented on src/components/Homepage/Nav/DropdownMenu.js:
```diff
@@ -0,0 +1,35 @@
+          <a href="#">Sofia</a>
+          <a href="#">Athens</a>
+          <a href="#">Istanbul</a>
+        </div>
```
1. Are those actual links that take you anywhere? If they're not, using a simple unordered list with list items inside will be a better fit.
2. Please create an array of objects that hold information for your cities(name, id) and render it instead of directly using them in the html

virzhinia-petkova-redshelf commented on src/components/Homepage/Nav/Nav.js:
```diff
@@ -0,0 +1,12 @@
+export function Nav(props) {
+  return (
+    <>
+      <nav>
```
not really sure nav is the right element to use here as you don't really use navigation in the dropdown

StefanAtanasov98 replied:
I imagined that when I click on a city which interests me in the dropdown, the view changes and on the page the user can see the weather for that specific city and have options like daily, weekly, monthly forecast and nothing else. From your comment I'm guessing that the view shouldn't change?

virzhinia-petkova-redshelf replied:
If you switch the view, you can leave it, it's fine. I like your idea. :)

virzhinia-petkova-redshelf commented on src/components/Homepage/Nav/Nav.js:
```diff
@@ -0,0 +1,12 @@
+
+export function Nav(props) {
+  return (
+    <>
```
this is not needed, fragments are used when you have more than one element at the same level and need something to wrap than so react works

virzhinia-petkova-redshelf commented on src/components/Homepage/Nav/Nav.js:
```diff
@@ -0,0 +1,12 @@
+import React from 'react';
+import { DropdownMenu } from './DropdownMenu.js';
+
+export function Nav(props) {
```
what is your idea with this component?

StefanAtanasov98 replied:
My idea is that Nav contains more things (buttons I guess) than just DropdownMenu, so that's why I think it should be a parent component for DropdownMenu as well as the other buttons

virzhinia-petkova-redshelf replied:
what buttons for example?

StefanAtanasov98 replied:
Maybe the ones for the forecast options like "Daily", "Weekly", etc.

virzhinia-petkova-redshelf replied:
okay

virzhinia-petkova-redshelf commented on src/components/Homepage/Nav/Nav.js:
```diff
@@ -0,0 +1,12 @@
+  return (
+    <>
+      <nav>
+        <DropdownMenu styles={props.styles} />
```
it's better to directly import styles to the component that uses them rather than passing them down

virzhinia-petkova-redshelf commented on src/styles/homepage.module.css:
```diff
@@ -0,0 +1,82 @@
+header {
```
please separate the styles for different components in different files, see https://create-react-app.dev/docs/adding-a-css-modules-stylesheet/

virzhinia-petkova-redshelf commented on src/styles/homepage.module.css:
```diff
@@ -0,0 +1,82 @@
+}
+
+.dropdownBtn {
+  background-color: #3498db;
```
please create variables for colors

virzhinia-petkova-redshelf commented on src/styles/homepage.module.css:
```diff
@@ -0,0 +1,82 @@
+
+.dropdownContent {
+  display: none;
+  /* position: absolute; */
```
please remove commented code

virzhinia-petkova-redshelf commented on src/components/Homepage/Nav/DropdownMenu.js:
```diff
@@ -0,0 +1,35 @@
+  };
+
+  return (
+    <>
```
I don't think you need this wrapper

virzhinia-petkova-redshelf reviewed (commented):
Overall good job. I left a few comments. 

Also, please provide brief a description about what work was done when opening the PR.

StefanAtanasov98 committed:
"Added reusable function for time consts, added Header and Footer components, added constants for magic strings, html refactored,
added state for cities and a list of cities for rendering, separated css files, created variables for css colors."

virzhinia-petkova-redshelf commented on src/components/Homepage/Nav/DropdownMenu.js:
```diff
@@ -0,0 +1,35 @@
+  { name: 'Sofia', id: Math.random() },
+  { name: 'Athens', id: Math.random() },
+  { name: 'Istanbul', id: Math.random() }
+];
```
you can just hard code something instead of using math.random - you'd need the id to be a constant value

virzhinia-petkova-redshelf commented on src/components/Homepage/Homepage.js:
```diff
@@ -0,0 +1,19 @@
+function HomePage(props) {
+  return (
+    <>
+      <Header />
```
would you want to reuse the header and footer for other pages without adding the component to each page? Where will you put it to be able to do that?

StefanAtanasov98 replied:
Oh, right, of course. If we're gonna use the same header and footer, I should put the components in App.jsx

virzhinia-petkova-redshelf commented on src/styles/DropdownMenu.module.css:
```diff
@@ -0,0 +1,28 @@
+:root {
```
if there a way to declare the variables in a way you can use them everywhere?

virzhinia-petkova-redshelf reviewed (commented):
please provide a description & screenshot for your PR

StefanAtanasov98 committed:
Refactoring, styling, adding functionality

Moved Header and Footer components to App.jsx, using a condition in JSX to show/hide dropdown content, separated css files, added a bit of styling
